File: agentic-adoption-scan/github.go
package main

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// RateLimitTracker tracks per-user rate limit state, keyed by a hash of the token.
// It can be shared across multiple GitHubClient instances (e.g. in the MCP HTTP server).
type RateLimitTracker struct {
	mu      sync.Mutex
	perUser map[string]*userRateState
}

type userRateState struct {
	searchLastCall  time.Time
	rateLimitRemain int
	rateLimitReset  time.Time
	lastSeen        time.Time
}

// NewRateLimitTracker creates a new RateLimitTracker.
func NewRateLimitTracker() *RateLimitTracker {
	return &RateLimitTracker{
		perUser: make(map[string]*userRateState),
	}
}

// getOrCreate returns the rate state for a token key, creating it if absent.
func (t *RateLimitTracker) getOrCreate(tokenKey string) *userRateState {
	t.mu.Lock()
	defer t.mu.Unlock()
	if s, ok := t.perUser[tokenKey]; ok {
		s.lastSeen = time.Now()
		return s
	}
	s := &userRateState{
		rateLimitRemain: 100, // assume budget until told otherwise
		lastSeen:        time.Now(),
	}
	t.perUser[tokenKey] = s
	return s
}

// CleanupStale removes entries that haven't been used within maxAge.
func (t *RateLimitTracker) CleanupStale(maxAge time.Duration) {
	t.mu.Lock()
	defer t.mu.Unlock()
	cutoff := time.Now().Add(-maxAge)
	for k, s := range t.perUser {
		if s.lastSeen.Before(cutoff) {
			delete(t.perUser, k)
		}
	}
}

// tokenKey returns a short, stable key for a token (sha256, first 16 hex chars).
func tokenKey(token string) string {
	h := sha256.Sum256([]byte(token))
	return fmt.Sprintf("%x", h[:8])
}

// GitHubClient wraps the GitHub API with rate limiting.
type GitHubClient struct {
	mu               sync.Mutex // local mutex, used when rateLimitTracker is nil
	logger           *log.Logger
	token            string
	httpClient       *http.Client
	searchMinDelay   time.Duration // minimum delay between code search calls
	rateLimitRemain  int           // local fallback when tracker is nil
	rateLimitReset   time.Time     // local fallback when tracker is nil
	searchLastCall   time.Time     // local fallback when tracker is nil
	rateLimitTracker *RateLimitTracker
}

// Repo represents a GitHub repository from the API.
type Repo struct {
	Name       string `json:"name"`
	FullName   string `json:"full_name"`
	Archived   bool   `json:"archived"`
	Visibility string `json:"visibility"`
	Language   string `json:"language"`
	PushedAt   string `json:"pushed_at"`
}

// SearchResult represents a code search hit.
type SearchResult struct {
	TotalCount int `json:"total_count"`
	Items      []struct {
		Name    string `json:"name"`
		Path    string `json:"path"`
		HTMLURL string `json:"html_url"`
	} `json:"items"`
}

// NewGitHubClient creates a client that reads the token from GH_TOKEN or GITHUB_TOKEN env vars.
func NewGitHubClient(logger *log.Logger) *GitHubClient {
	token := os.Getenv("GH_TOKEN")
	if token == "" {
		token = os.Getenv("GITHUB_TOKEN")
	}
	return NewGitHubClientWithToken(token, logger)
}

// NewGitHubClientWithToken creates a client using the provided token.
func NewGitHubClientWithToken(token string, logger *log.Logger) *GitHubClient {
	return &GitHubClient{
		logger:          logger,
		token:           token,
		httpClient:      &http.Client{Timeout: 30 * time.Second},
		searchMinDelay:  2100 * time.Millisecond, // stay under 30 req/min for code search
		rateLimitRemain: 100,                      // assume we have budget until told otherwise
	}
}

const githubAPIBase = "https://api.github.com"

// ghAPI calls the GitHub REST API and returns the raw JSON body.
// It handles rate limiting (HTTP 403/429) and retries with backoff.
func (c *GitHubClient) ghAPI(method, endpoint string) ([]byte, *http.Response, error) {
	c.waitForRateLimit()

	// Build the full URL. Endpoints from callers start with "/".
	rawURL := githubAPIBase + endpoint

	var lastErr error
	for attempt := 0; attempt < 4; attempt++ {
		if attempt > 0 {
			backoff := time.Duration(1<<uint(attempt)) * time.Second
			jitter := time.Duration(rand.Intn(500)) * time.Millisecond
			c.logger.Printf("Rate limited, retrying in %v (attempt %d)", backoff+jitter, attempt+1)
			time.Sleep(backoff + jitter)
		}

		req, err := http.NewRequest(method, rawURL, nil)
		if err != nil {
			return nil, nil, fmt.Errorf("building request: %w", err)
		}

		req.Header.Set("Accept", "application/vnd.github+json")
		req.Header.Set("X-GitHub-Api-Version", "2022-11-28")
		if c.token != "" {
			req.Header.Set("Authorization", "Bearer "+c.token)
		}

		resp, err := c.httpClient.Do(req)
		if err != nil {
			lastErr = fmt.Errorf("http request: %w", err)
			continue
		}

		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			lastErr = fmt.Errorf("reading response body: %w", err)
			continue
		}

		// Update rate limit state from response headers
		c.updateRateLimit(resp)

		// Handle rate limiting
		if resp.StatusCode == 403 || resp.StatusCode == 429 {
			lastErr = fmt.Errorf("rate limited (HTTP %d): %s", resp.StatusCode, string(body))
			continue
		}

		if resp.StatusCode >= 400 {
			return nil, resp, fmt.Errorf("github api error (HTTP %d): %s", resp.StatusCode, string(body))
		}

		return body, resp, nil
	}
	return nil, nil, fmt.Errorf("exhausted retries: %w", lastErr)
}

// ghAPIPaginated fetches all pages of a paginated endpoint, using the Link header.
func (c *GitHubClient) ghAPIPaginated(endpoint string) ([]json.RawMessage, error) {
	var allItems []json.RawMessage

	// Build initial URL with per_page=100
	sep := "?"
	if strings.Contains(endpoint, "?") {
		sep = "&"
	}
	nextURL := githubAPIBase + endpoint + sep + "per_page=100"

	for nextURL != "" {
		c.waitForRateLimit()

		req, err := http.NewRequest("GET", nextURL, nil)
		if err != nil {
			return nil, fmt.Errorf("building request: %w", err)
		}
		req.Header.Set("Accept", "application/vnd.github+json")
		req.Header.Set("X-GitHub-Api-Version", "2022-11-28")
		if c.token != "" {
			req.Header.Set("Authorization", "Bearer "+c.token)
		}

		var lastErr error
		var body []byte
		var resp *http.Response

		for attempt := 0; attempt < 4; attempt++ {
			if attempt > 0 {
				backoff := time.Duration(1<<uint(attempt)) * time.Second
				jitter := time.Duration(rand.Intn(500)) * time.Millisecond
				c.logger.Printf("Rate limited, retrying in %v (attempt %d)", backoff+jitter, attempt+1)
				time.Sleep(backoff + jitter)
			}

			resp, err = c.httpClient.Do(req)
			if err != nil {
				lastErr = fmt.Errorf("http request: %w", err)
				continue
			}

			body, err = io.ReadAll(resp.Body)
			resp.Body.Close()
			if err != nil {
				lastErr = fmt.Errorf("reading response body: %w", err)
				continue
			}

			c.updateRateLimit(resp)

			if resp.StatusCode == 403 || resp.StatusCode == 429 {
				lastErr = fmt.Errorf("rate limited (HTTP %d): %s", resp.StatusCode, string(body))
				continue
			}

			if resp.StatusCode >= 400 {
				return nil, fmt.Errorf("github api error (HTTP %d): %s", resp.StatusCode, string(body))
			}

			lastErr = nil
			break
		}
		if lastErr != nil {
			return nil, fmt.Errorf("exhausted retries: %w", lastErr)
		}

		var items []json.RawMessage
		if err := json.Unmarshal(body, &items); err != nil {
			return nil, fmt.Errorf("failed to parse paginated response: %w", err)
		}

		allItems = append(allItems, items...)

		// Parse Link header for next page
		nextURL = parseLinkNext(resp.Header.Get("Link"))
	}

	return allItems, nil
}

// parseLinkNext extracts the URL for rel="next" from a Link header value.
func parseLinkNext(header string) string {
	if header == "" {
		return ""
	}
	// Link: <https://api.github.com/...?page=2>; rel="next", <...>; rel="last"
	for _, part := range strings.Split(header, ",") {
		part = strings.TrimSpace(part)
		segments := strings.Split(part, ";")
		if len(segments) < 2 {
			continue
		}
		urlPart := strings.TrimSpace(segments[0])
		relPart := strings.TrimSpace(segments[1])
		if relPart == `rel="next"` {
			// Strip angle brackets
			urlPart = strings.TrimPrefix(urlPart, "<")
			urlPart = strings.TrimSuffix(urlPart, ">")
			return urlPart
		}
	}
	return ""
}

// ListOrgRepos lists all repositories in an organization.
func (c *GitHubClient) ListOrgRepos(org string) ([]Repo, error) {
	c.logger.Printf("Listing repos for org: %s", org)

	raw, err := c.ghAPIPaginated(fmt.Sprintf("/orgs/%s/repos?sort=pushed&direction=desc", org))
	if err != nil {
		return nil, fmt.Errorf("listing org repos: %w", err)
	}

	var repos []Repo
	for _, item := range raw {
		var r Repo
		if err := json.Unmarshal(item, &r); err != nil {
			c.logger.Printf("Warning: failed to parse repo: %v", err)
			continue
		}
		repos = append(repos, r)
	}

	c.logger.Printf("Found %d repos in %s", len(repos), org)
	return repos, nil
}

// CheckPathExists checks if a file or directory exists in a repo.
// Returns (exists, isDir, error).
func (c *GitHubClient) CheckPathExists(owner, repo, path string) (bool, bool, error) {
	c.logger.Printf("Checking path: %s/%s/%s", owner, repo, path)

	out, resp, err := c.ghAPI("GET", fmt.Sprintf("/repos/%s/%s/contents/%s", owner, repo, path))
	if err != nil {
		if resp != nil && resp.StatusCode == 404 {
			return false, false, nil
		}
		errStr := err.Error()
		if strings.Contains(errStr, "404") || strings.Contains(errStr, "Not Found") {
			return false, false, nil
		}
		return false, false, err
	}

	// If it's a directory, the response is an array
	trimmed := strings.TrimSpace(string(out))
	if len(trimmed) > 0 && trimmed[0] == '[' {
		return true, true, nil
	}

	// Single file response
	var content struct {
		Type string `json:"type"`
	}
	if err := json.Unmarshal(out, &content); err == nil {
		return true, content.Type == "dir", nil
	}

	return true, false, nil
}

// SearchCode searches for code patterns in a specific repo using the Code Search API.
func (c *GitHubClient) SearchCode(org, repo, query string) (*SearchResult, error) {
	c.throttleSearch()
	c.logger.Printf("Code search in %s/%s: %s", org, repo, query)

	fullQuery := fmt.Sprintf("%s repo:%s/%s", query, org, repo)
	encodedQuery := url.QueryEscape(fullQuery)

	out, _, err := c.ghAPI("GET", "/search/code?q="+encodedQuery)
	if err != nil {
		return nil, err
	}

	var result SearchResult
	if err := json.Unmarshal(out, &result); err != nil {
		return nil, fmt.Errorf("parsing search results: %w", err)
	}

	return &result, nil
}

// SearchCodeInWorkflows searches for patterns specifically in workflow files.
func (c *GitHubClient) SearchCodeInWorkflows(org, repo, query string) (*SearchResult, error) {
	return c.SearchCode(org, repo, fmt.Sprintf("%s path:.github/workflows", query))
}

// GetFileContent fetches the raw content of a file from a repo.
func (c *GitHubClient) GetFileContent(owner, repo, path string) ([]byte, error) {
	c.logger.Printf("Fetching content: %s/%s/%s", owner, repo, path)

	endpoint := fmt.Sprintf("/repos/%s/%s/contents/%s", owner, repo, path)
	rawURL := githubAPIBase + endpoint

	req, err := http.NewRequest("GET", rawURL, nil)
	if err != nil {
		return nil, fmt.Errorf("building request: %w", err)
	}

	req.Header.Set("Accept", "application/vnd.github.raw+json")
	req.Header.Set("X-GitHub-Api-Version", "2022-11-28")
	if c.token != "" {
		req.Header.Set("Authorization", "Bearer "+c.token)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("fetching file content: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("reading file content: %w", err)
	}

	c.updateRateLimit(resp)

	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("fetching file content (HTTP %d): %s", resp.StatusCode, string(body))
	}

	return body, nil
}

// throttleSearch enforces the code search rate limit (30 req/min).
func (c *GitHubClient) throttleSearch() {
	if c.rateLimitTracker != nil {
		key := tokenKey(c.token)
		state := c.rateLimitTracker.getOrCreate(key)
		c.rateLimitTracker.mu.Lock()
		defer c.rateLimitTracker.mu.Unlock()

		elapsed := time.Since(state.searchLastCall)
		if elapsed < c.searchMinDelay {
			wait := c.searchMinDelay - elapsed
			c.logger.Printf("Throttling search: waiting %v", wait)
			time.Sleep(wait)
		}
		state.searchLastCall = time.Now()
		return
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	elapsed := time.Since(c.searchLastCall)
	if elapsed < c.searchMinDelay {
		wait := c.searchMinDelay - elapsed
		c.logger.Printf("Throttling search: waiting %v", wait)
		time.Sleep(wait)
	}
	c.searchLastCall = time.Now()
}

// waitForRateLimit checks if we're close to the rate limit and waits if needed.
func (c *GitHubClient) waitForRateLimit() {
	if c.rateLimitTracker != nil {
		key := tokenKey(c.token)
		state := c.rateLimitTracker.getOrCreate(key)
		c.rateLimitTracker.mu.Lock()
		defer c.rateLimitTracker.mu.Unlock()

		if state.rateLimitRemain < 10 && time.Now().Before(state.rateLimitReset) {
			wait := time.Until(state.rateLimitReset) + time.Second
			c.logger.Printf("Rate limit low (%d remaining), waiting %v until reset", state.rateLimitRemain, wait)
			time.Sleep(wait)
		}
		return
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	if c.rateLimitRemain < 10 && time.Now().Before(c.rateLimitReset) {
		wait := time.Until(c.rateLimitReset) + time.Second
		c.logger.Printf("Rate limit low (%d remaining), waiting %v until reset", c.rateLimitRemain, wait)
		time.Sleep(wait)
	}
}

// updateRateLimit parses X-RateLimit-Remaining and X-RateLimit-Reset from HTTP response headers.
func (c *GitHubClient) updateRateLimit(resp *http.Response) {
	remainStr := resp.Header.Get("X-RateLimit-Remaining")
	resetStr := resp.Header.Get("X-RateLimit-Reset")

	if remainStr == "" && resetStr == "" {
		return
	}

	remain, err := strconv.Atoi(remainStr)
	if err != nil {
		remain = -1
	}

	var reset time.Time
	if resetUnix, err := strconv.ParseInt(resetStr, 10, 64); err == nil {
		reset = time.Unix(resetUnix, 0)
	}

	if c.rateLimitTracker != nil {
		key := tokenKey(c.token)
		state := c.rateLimitTracker.getOrCreate(key)
		c.rateLimitTracker.mu.Lock()
		defer c.rateLimitTracker.mu.Unlock()
		if remain >= 0 {
			state.rateLimitRemain = remain
		}
		if !reset.IsZero() {
			state.rateLimitReset = reset
		}
		return
	}

	c.mu.Lock()
	defer c.mu.Unlock()
	if remain >= 0 {
		c.rateLimitRemain = remain
	}
	if !reset.IsZero() {
		c.rateLimitReset = reset
	}
}

File: agentic_adoption_scan/github.py
"""GitHub API client with per-user rate limiting and retry logic.

Port of agentic-adoption-scan/github.go.
"""

from __future__ import annotations

import hashlib
import logging
import os
import random
import threading
import time
import urllib.parse
from dataclasses import dataclass, field
from typing import Optional

import httpx

logger = logging.getLogger(__name__)

GITHUB_API_BASE = "https://api.github.com"
_SEARCH_MIN_DELAY = 2.1  # seconds — stay under 30 req/min for code search


# ---------------------------------------------------------------------------
# Rate-limit state
# ---------------------------------------------------------------------------


class UserRateState:
    """Per-user rate limit state (all fields protected by lock)."""

    def __init__(self) -> None:
        self.lock = threading.Lock()
        self.search_last_call: float = 0.0
        self.rate_limit_remain: int = 100  # assume budget until told otherwise
        self.rate_limit_reset: float = 0.0
        self.last_seen: float = time.monotonic()


class RateLimitTracker:
    """Per-user rate limit tracker, keyed by sha256(token)[:16].

    Designed to be shared across multiple GitHubClient instances (e.g. in the
    MCP HTTP server where each request carries its own Bearer token).
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._per_user: dict[str, UserRateState] = {}

    def get_or_create(self, token_key: str) -> UserRateState:
        with self._lock:
            state = self._per_user.get(token_key)
            if state is not None:
                state.last_seen = time.monotonic()
                return state
            state = UserRateState()
            self._per_user[token_key] = state
            return state

    def cleanup_stale(self, max_age: float) -> None:
        """Remove entries that haven't been used within max_age seconds."""
        cutoff = time.monotonic() - max_age
        with self._lock:
            stale = [k for k, s in self._per_user.items() if s.last_seen < cutoff]
            for k in stale:
                del self._per_user[k]

    def start_cleanup(
        self,
        interval: float,
        max_age: float,
        stop_event: threading.Event,
    ) -> None:
        """Run periodic cleanup in a background daemon thread."""

        def _loop() -> None:
            while not stop_event.wait(timeout=interval):
                self.cleanup_stale(max_age)

        t = threading.Thread(target=_loop, daemon=True, name="rate-limit-cleanup")
        t.start()


# ---------------------------------------------------------------------------
# Data models
# ---------------------------------------------------------------------------


@dataclass
class Repo:
    name: str
    full_name: str
    archived: bool
    visibility: str
    language: str
    pushed_at: str


@dataclass
class SearchResult:
    total_count: int
    items: list[dict]  # each dict has "name", "path", "html_url"


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _token_key(token: str) -> str:
    """Return sha256(token)[:16] — stable, short key for a token."""
    return hashlib.sha256(token.encode()).hexdigest()[:16]


def _parse_link_next(header: str) -> str:
    """Extract the URL for rel="next" from a Link header value.

    Example header:
        <https://api.github.com/...?page=2>; rel="next", <...>; rel="last"
    """
    if not header:
        return ""
    for part in header.split(","):
        part = part.strip()
        segments = part.split(";")
        if len(segments) < 2:
            continue
        url_part = segments[0].strip()
        rel_part = segments[1].strip()
        if rel_part == 'rel="next"':
            # Strip angle brackets
            url_part = url_part.lstrip("<").rstrip(">")
            return url_part
    return ""


# ---------------------------------------------------------------------------
# GitHubClient
# ---------------------------------------------------------------------------


class GitHubClient:
    """Sync httpx client wrapping the GitHub REST API.

    Handles per-user rate limiting, exponential-backoff retries, and
    pagination.  Use ``from_env()`` for the common single-token case, or
    pass an explicit token for per-request auth (MCP HTTP server).
    """

    def __init__(
        self,
        token: str,
        log: Optional[logging.Logger] = None,
        tracker: Optional[RateLimitTracker] = None,
    ) -> None:
        self._token = token
        self._log = log or logger
        self._tracker = tracker
        self._client = httpx.Client(timeout=30.0)

        # Local fallback state (used when tracker is None)
        self._local_lock = threading.Lock()
        self._local_rate_limit_remain: int = 100
        self._local_rate_limit_reset: float = 0.0
        self._local_search_last_call: float = 0.0

    @classmethod
    def from_env(cls, log: Optional[logging.Logger] = None) -> "GitHubClient":
        """Create a client by reading GH_TOKEN then GITHUB_TOKEN from the environment."""
        token = os.environ.get("GH_TOKEN") or os.environ.get("GITHUB_TOKEN") or ""
        return cls(token=token, log=log)

    # ------------------------------------------------------------------
    # Core HTTP
    # ------------------------------------------------------------------

    def api(
        self,
        method: str,
        endpoint: str,
        accept: str = "",
    ) -> tuple[bytes, httpx.Response]:
        """Call the GitHub REST API and return (body, response).

        Raises on unrecoverable errors or after exhausting all retries.
        """
        self._wait_for_rate_limit()

        raw_url = GITHUB_API_BASE + endpoint
        if not accept:
            accept = "application/vnd.github+json"

        last_exc: Optional[Exception] = None
        for attempt in range(4):
            if attempt > 0:
                backoff = (1 << attempt)  # 2, 4, 8 seconds
                jitter = random.randint(0, 500) / 1000.0
                wait = backoff + jitter
                self._log.warning(
                    "Rate limited, retrying in %.2fs (attempt %d)", wait, attempt + 1
                )
                time.sleep(wait)

            headers: dict[str, str] = {
                "Accept": accept,
                "X-GitHub-Api-Version": "2022-11-28",
            }
            if self._token:
                headers["Authorization"] = "Bearer " + self._token

            try:
                resp = self._client.request(method, raw_url, headers=headers)
            except httpx.RequestError as exc:
                last_exc = exc
                continue

            body = resp.content
            self._update_rate_limit(resp)

            if resp.status_code in (403, 429):
                last_exc = RuntimeError(
                    f"rate limited (HTTP {resp.status_code}): {body.decode(errors='replace')}"
                )
                continue

            if resp.status_code >= 400:
                raise RuntimeError(
                    f"github api error (HTTP {resp.status_code}): {body.decode(errors='replace')}"
                )

            return body, resp

        raise RuntimeError(f"exhausted retries: {last_exc}") from last_exc

    def api_paginated(self, endpoint: str) -> list[dict]:
        """Fetch all pages of a paginated endpoint using the Link header."""
        import json as _json

        sep = "&" if "?" in endpoint else "?"
        current = endpoint + sep + "per_page=100"
        all_items: list[dict] = []

        while current:
            body, resp = self.api("GET", current)
            items: list[dict] = _json.loads(body)
            all_items.extend(items)

            next_url = _parse_link_next(resp.headers.get("link", ""))
            if not next_url:
                break
            # Link header returns full URLs; strip the base to get an endpoint
            current = next_url.removeprefix(GITHUB_API_BASE)

        return all_items

    # ------------------------------------------------------------------
    # Public API methods
    # ------------------------------------------------------------------

    def list_org_repos(self, org: str) -> list[Repo]:
        self._log.info("Listing repos for org: %s", org)
        raw = self.api_paginated(f"/orgs/{org}/repos?sort=pushed&direction=desc")
        repos: list[Repo] = []
        for item in raw:
            try:
                repos.append(
                    Repo(
                        name=item.get("name", ""),
                        full_name=item.get("full_name", ""),
                        archived=bool(item.get("archived", False)),
                        visibility=item.get("visibility", ""),
                        language=item.get("language") or "",
                        pushed_at=item.get("pushed_at") or "",
                    )
                )
            except Exception as exc:  # noqa: BLE001
                self._log.warning("Failed to parse repo: %s", exc)
        self._log.info("Found %d repos in %s", len(repos), org)
        return repos

    def check_path_exists(self, owner: str, repo: str, path: str) -> tuple[bool, bool]:
        """Check if a path exists in a repo. Returns (exists, is_dir)."""
        self._log.debug("Checking path: %s/%s/%s", owner, repo, path)
        endpoint = f"/repos/{owner}/{repo}/contents/{path}"

        try:
            body, _ = self.api("GET", endpoint)
        except RuntimeError as exc:
            msg = str(exc)
            if "404" in msg or "Not Found" in msg:
                return False, False
            raise

        # Directory responses are JSON arrays
        trimmed = body.lstrip()
        if trimmed and trimmed[0:1] == b"[":
            return True, True

        # Single file response
        import json as _json

        try:
            content = _json.loads(body)
            return True, content.get("type") == "dir"
        except Exception:  # noqa: BLE001
            return True, False

    def search_code(self, org: str, repo: str, query: str) -> SearchResult:
        """Search for code in a specific repo using the Code Search API."""
        self._throttle_search()
        self._log.debug("Code search in %s/%s: %s", org, repo, query)

        import json as _json

        full_query = f"{query} repo:{org}/{repo}"
        encoded = urllib.parse.quote(full_query)

        body, _ = self.api("GET", f"/search/code?q={encoded}")
        data = _json.loads(body)

        items = [
            {
                "name": i.get("name", ""),
                "path": i.get("path", ""),
                "html_url": i.get("html_url", ""),
            }
            for i in data.get("items", [])
        ]
        return SearchResult(total_count=data.get("total_count", 0), items=items)

    def search_code_in_workflows(self, org: str, repo: str, query: str) -> SearchResult:
        """Search for patterns specifically inside .github/workflows."""
        return self.search_code(org, repo, f"{query} path:.github/workflows")

    def get_file_content(self, owner: str, repo: str, path: str) -> bytes:
        """Fetch the raw content of a file from a repo."""
        self._log.debug("Fetching content: %s/%s/%s", owner, repo, path)
        endpoint = f"/repos/{owner}/{repo}/contents/{path}"
        body, _ = self.api("GET", endpoint, accept="application/vnd.github.raw+json")
        return body

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _throttle_search(self) -> None:
        """Enforce the code search rate limit (≥2100 ms between calls)."""
        if self._tracker is not None:
            key = _token_key(self._token)
            state = self._tracker.get_or_create(key)

            # Read state under per-user lock, sleep OUTSIDE the lock
            with state.lock:
                elapsed = time.monotonic() - state.search_last_call

            if elapsed < _SEARCH_MIN_DELAY:
                wait = _SEARCH_MIN_DELAY - elapsed
                self._log.debug("Throttling search: waiting %.3fs", wait)
                time.sleep(wait)

            with state.lock:
                state.search_last_call = time.monotonic()
            return

        # Local fallback
        with self._local_lock:
            elapsed = time.monotonic() - self._local_search_last_call

        if elapsed < _SEARCH_MIN_DELAY:
            wait = _SEARCH_MIN_DELAY - elapsed
            self._log.debug("Throttling search: waiting %.3fs", wait)
            time.sleep(wait)

        with self._local_lock:
            self._local_search_last_call = time.monotonic()

    def _wait_for_rate_limit(self) -> None:
        """Sleep until the rate limit resets if remaining requests are low."""
        if self._tracker is not None:
            key = _token_key(self._token)
            state = self._tracker.get_or_create(key)

            # Read state under per-user lock, sleep OUTSIDE the lock
            with state.lock:
                should_wait = (
                    state.rate_limit_remain < 10
                    and state.rate_limit_reset > time.time()
                )
                wait = (state.rate_limit_reset - time.time() + 1.0) if should_wait else 0.0

            if should_wait:
                self._log.warning("Rate limit low, waiting %.1fs until reset", wait)
                time.sleep(wait)
            return

        # Local fallback
        with self._local_lock:
            should_wait = (
                self._local_rate_limit_remain < 10
                and self._local_rate_limit_reset > time.time()
            )
            wait = (
                (self._local_rate_limit_reset - time.time() + 1.0) if should_wait else 0.0
            )

        if should_wait:
            self._log.warning(
                "Rate limit low (%d remaining), waiting %.1fs until reset",
                self._local_rate_limit_remain,
                wait,
            )
            time.sleep(wait)

    def _update_rate_limit(self, resp: httpx.Response) -> None:
        """Parse X-RateLimit-Remaining and X-RateLimit-Reset response headers."""
        remain_str = resp.headers.get("x-ratelimit-remaining", "")
        reset_str = resp.headers.get("x-ratelimit-reset", "")

        if not remain_str and not reset_str:
            return

        remain = -1
        if remain_str:
            try:
                remain = int(remain_str)
            except ValueError:
                pass

        reset = 0.0
        if reset_str:
            try:
                reset = float(reset_str)
            except ValueError:
                pass

        if self._tracker is not None:
            key = _token_key(self._token)
            state = self._tracker.get_or_create(key)
            with state.lock:
                if remain >= 0:
                    state.rate_limit_remain = remain
                if reset:
                    state.rate_limit_reset = reset
            return

        with self._local_lock:
            if remain >= 0:
                self._local_rate_limit_remain = remain
            if reset:
                self._local_rate_limit_reset = reset
